crates/swc_ecma_transforms_compat/src/bugfixes/safari_id_destructuring_collision_in_function_expression.rs

```diff
@@ -0,0 +1,115 @@
+use swc_atoms::JsWord;
+use swc_common::{
+    collections::{AHashMap, AHashSet},
+    Span,
+};
+use swc_ecma_ast::*;
+use swc_ecma_transforms_base::hygiene::rename;
+use swc_ecma_visit::{as_folder, noop_visit_mut_type, Fold, VisitMut, VisitMutWith};
+
+pub fn safari_id_destructuring_collision_in_function_expression() -> impl Fold + VisitMut {
+    as_folder(SafariIdDestructuringCollisionInFunctionExpression::default())
+}
+
+#[derive(Default, Clone)]
+struct SafariIdDestructuringCollisionInFunctionExpression {
+    fn_expr_name: JsWord,
+    destructured_id_span: Option<Span>,
+    other_ident_syms: AHashSet<JsWord>,
+    in_body: bool,
+}
+
+impl VisitMut for SafariIdDestructuringCollisionInFunctionExpression {
+    noop_visit_mut_type!();
+
+    fn visit_mut_binding_ident(&mut self, binding_ident: &mut BindingIdent) {
+        if !self.in_body && self.fn_expr_name.eq(&binding_ident.id.sym) {
+            self.destructured_id_span = Some(binding_ident.id.span);
+        } else {
+            self.other_ident_syms.insert(binding_ident.id.sym.clone());
+        }
+    }
+
+    fn visit_mut_ident(&mut self, ident: &mut Ident) {
+        if self.in_body && !self.fn_expr_name.eq(&ident.sym) {
+            self.other_ident_syms.insert(ident.sym.clone());
+        }
+    }
+
+    fn visit_mut_param(&mut self, param: &mut Param) {
+        param.visit_mut_children_with(self);
+    }
+
+    fn visit_mut_block_stmt(&mut self, block: &mut BlockStmt) {
+        block.visit_mut_children_with(self);
+    }
+
+    fn visit_mut_fn_expr(&mut self, n: &mut FnExpr) {
+        if let Some(ident) = &n.ident {
+            self.fn_expr_name = ident.sym.clone();
+            n.function.params.visit_mut_children_with(self);
+            self.in_body = true;
+            n.function.body.visit_mut_children_with(self);
+            self.in_body = false;
+            if let Some(id_span) = &self.destructured_id_span {
+                let mut rename_map = AHashMap::default();
+                let new_id: JsWord = {
+                    let mut id_value: JsWord = format!("_{}", self.fn_expr_name).into();
```

Comment:
I think this will break code which declares a variable from the outside of a function and assign to it in the function.

Reply:
In such case, this will only rename the assignment in the function.

Reply:
> code which declares a variable from the outside of a function and assign to it in the function.

like below?:

```js
let _a;
(function a([a]) {
  _a = 3;
});
```

Reply:
Yes, that's the case I mentioned.

Reply:
In current implementation, it works fine!

https://github.com/swc-project/swc/blob/bef2dd9f4c95c30f7fb2540e735ff9c194e335a2/crates/swc_ecma_transforms_compat/src/bugfixes/safari_id_destructuring_collision_in_function_expression.rs#L108-L120

Reply:
Can you add an integration test for 

```js
var a = 1;
var _a = 2;


const obj = {
     a: function () {
         a = 2;
         _a = 3;
     }
}

```

?